### DISCUSSION TITLE
Random freezes in dev mode

### DISCUSSION THREAD

**Discussion by dolzenko** [puma/puma#3352] [Q&A]
My Rails 5 (big) app randomly freezes in development mode, attached thread dump. Hopefully someone can help me with the next steps in debugging this
[hang.txt](https://github.com/puma/puma/files/14697254/hang.txt).
I have tried comparing this to the threaddump of not hanging app and it seems the main difference is that "puma server" thread is in the runnable state (compared to TIMED_WAITING when hanging, but not sure how to check what is it waiting on).

**Puma config:**

default, on start puma prints:

```
Puma starting in single mode...
* Version 4.3.12 (jruby 9.3.11.0 - ruby 2.6.8), codename: Mysterious Traveller
* Min threads: 0, max threads: 16
* Environment: development
* Listening on tcp://0.0.0.0:3000
```

**Desktop:**
 - OS: Linux
 - Puma Version 4.3.12

**Comment by dentarg** [comment]
Try Puma 6?

the issue is most likely in your app and impossible to debug with access to it

**Comment by dolzenko** [comment]
> Try Puma 6?
> 
> the issue is most likely in your app and impossible to debug with access to it

upgrade is certainly on our list of things to try, but this is a big app with many interlocking dependencies, not sure how feasible is the upgrade at the moment :( It's also "the average Rails app", nothing fancy,

**Comment by nateberkopec** [comment]
Jruby stack traces are not my strong suit but this looks like waiting on a lock in your logger (log4r).
